go/channelling/turnservice_manager.go

```diff
@@ -0,0 +1,124 @@
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package channelling
+
+import (
+	"log"
+	"sync"
+	"time"
+
+	"github.com/strukturag/spreed-turnservicecli/turnservicecli"
+)
+
+type TURNServiceManager interface {
+	TurnDataCreator
+}
+
+type turnServiceManager struct {
+	sync.Mutex
+	pleaders map[uint64]Sender
+
+	uri         string
+	accessToken string
+	clientID    string
+	turnService *turnservicecli.TURNService
+}
+
+func NewTURNServiceManager(uri string, accessToken string, clientID string) TURNServiceManager {
+	turnService := turnservicecli.NewTURNService(uri, 0, nil)
+	mgr := &turnServiceManager{
+		uri:         uri,
+		accessToken: accessToken,
+		clientID:    clientID,
+
+		turnService: turnService,
+		pleaders:    make(map[uint64]Sender),
+	}
+
+	turnService.Open(accessToken, clientID, "")
+	turnService.BindOnCredentials(mgr.onCredentials)
+	log.Println("Fetching TURN credentials from service")
+	go func() {
+		//time.Sleep(10000 * time.Millisecond)
+		turnService.Autorefresh(true)
+	}()
+	// Wait a bit, to give TURN service some time to populate credentials, so
+	// we avoid to have send them as an update for fast reconnecting clients.
+	time.Sleep(500 * time.Millisecond)
+	if mgr.turnService.Credentials(false) == nil {
+		log.Println("No TURN credentials from service on startup - extra traffic for clients connecting before credentials have been received")
+	}
+
+	return mgr
+}
+
+func (mgr *turnServiceManager) CreateTurnData(sender Sender, session *Session) *DataTurn {
+	credentials := mgr.turnService.Credentials(false)
+	turn, err := mgr.turnData(credentials)
+	if err != nil || turn.Ttl == 0 {
+		// When no data was return from service, refresh quickly.
+		mgr.Lock()
+		mgr.pleaders[sender.Index()] = sender
+		mgr.Unlock()
+
+		// Have client come back early.
+		turn.Ttl = 300
+	}
+
+	return turn
+}
+
+func (mgr *turnServiceManager) turnData(credentials *turnservicecli.CachedCredentialsData) (*DataTurn, error) {
+	turn := &DataTurn{}
+	if credentials != nil {
+		turn.Username = credentials.Turn.Username
+		turn.Password = credentials.Turn.Password
+		turn.Servers = credentials.Turn.Servers
+		turn.Ttl = int(credentials.TTL())
+		turn.GeoURI = credentials.Turn.GeoURI
+	} else {
+		turn.Ttl = 0
+	}
+
+	return turn, nil
+}
+
+func (mgr *turnServiceManager) onCredentials(credentials *turnservicecli.CachedCredentialsData, err error) {
+	if err != nil {
+		log.Printf("TURN credentials service error: %s\n", err.Error())
+		return
+	}
+
+	log.Println("Received TURN credentials from service", credentials.Turn.Username)
+
+	mgr.Lock()
+	for _, sender := range mgr.pleaders {
+		if turn, err := mgr.turnData(credentials); err == nil && turn.Ttl > 0 {
```

Comment:
The `turn.Ttl > 0` is unnecessary as this method is only called when new TURN credentials are received (which always should be valid initially).